### PR TITLE
Add Socket option helper methods.

### PR DESCRIPTION
**Pull request by technorama**

### CODE REVIEW

**Comment by technorama** [comment]
@jhass This PR adds specs for non-integer setsockopt usage (linger) you requested in #1278.

**Comment by jhass** [comment]
I noticed, thank you!

File: src/io/file_descriptor_io.cr [hunk 1]
```diff
@@ -258,6 +259,8 @@ class FileDescriptorIO
     if event = @event
       event.free
     end
+
+    raise err if err
```
Comment (lbguilherme) [inline src/io/file_descriptor_io.cr]: If there is an error trying to close the file, should we really mark the file as closed?

**Comment by technorama** [comment]
The linux man page says close return -1 on the following errors:
* EBADF fd isn't a valid open file descriptor.
* EINTR The close() call was interrupted by a signal; see signal(7).
* EIO An I/O error occurred.

EINTR isn't really possible on Linux.  Source: https://lwn.net/Articles/576478/

Neither of the other 2 errors are recoverable, so I believe the object should be marked as closed so that no other IO operations take place.

If there is a recoverable error the close operation should be retried in IO#close and not in individual applications.

**Comment by bnoordhuis** [comment]
> EINTR isn't really possible on Linux.

This is not quite true if you count older kernels or other UNIX platforms.  POSIX also allows returning EINPROGRESS and musl libc does that, for example.

I'd just ignore EINTR and EINPROGRESS.  The only platform I know of where ignoring EINTR on close() results in a file descriptor leak is HP-UX and I doubt anyone cares about that.

**Comment by asterite** [comment]
@technorama In which cases do we need the linger and other options? We really don't want to add every possible method to the API if it's not going to be used.

**Comment by technorama** [comment]
Every option in this PR (except linger) is one that I use.  Keepalive timeouts and intervals are similar to read_timeout but serve different purposes.  Keepalives notify you within a certain amount of time if the connection to the peer is unreachable.  

Here's a scenario:
```
begin
  io.tcp_keepalive_idle = 10
  io.read_fully(frame)
rescue ex
  # connect to a different peer (p2p)
end
``` 

`read_timeout` can't be used as a replacement for keepalives as the peer may not send data for extended periods of time but is still connected.

`shutdown_read` and `shutdown_write` are ABSOLUTELY necessary for bidirectional protocols.  It's the reliable way to avoid linger.  `shutdown_read` can also be used in forked applications to stop `accept` in children and allow for graceful shutdowns.

`linger` was added to satisfy a request by @jhass in #1278.  There aren't any other non-integer options that I'm familiar with and feel comfortable adding.

Please note there are many other socket options that weren't added because I don't use them, they are platform specific, or exotic.  This PR is my curated list of socket options (except linger).

**Comment by asterite** [comment]
Hm, and it seems all of these are also available in Ruby, but in different forms (for example [Socket::Options](http://ruby-doc.org/stdlib-1.9.3/libdoc/socket/rdoc/Socket/Option.html#method-c-linger)). I'll ask @waj, but we'll probably merge this.

Could you rebase it please?

**Pull request merged**